sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusClientFactory.cs

```diff
@@ -0,0 +1,226 @@
+                entityPath,
+                new ServiceBusReceiverOptions
+            {
+                PrefetchCount = _options.PrefetchCount
+            }));
+        }
+
+        private ServiceBusProcessor GetOrAddProcessor(string entityPath, string connection)
+        {
+            ServiceBusClient client = GetOrCreateClient(connection);
+
+            return _processorCache.GetOrAdd(entityPath, (_) =>
+            {
+                ServiceBusProcessor processor;
+                if (ServiceBusEntityPathHelper.ParseEntityType(entityPath) == EntityType.Topic)
+                {
+                    // entityPath for a subscription is "{TopicName}/Subscriptions/{SubscriptionName}"
+                    ServiceBusEntityPathHelper.ParseTopicAndSubscription(entityPath, out string topic, out string subscription);
+                    processor = client.CreateProcessor(topic, subscription, _options.ToProcessorOptions());
+                }
+                else
+                {
+                    // entityPath for a queue is "{QueueName}"
+                    processor = client.CreateProcessor(entityPath, _options.ToProcessorOptions());
+                }
+                processor.ProcessErrorAsync += _options.ExceptionReceivedHandler;
+                return processor;
+            });
+        }
+
+        private ServiceBusSessionProcessor GetOrAddSessionProcessor(string entityPath, string connection)
+        {
+            ServiceBusClient client = GetOrCreateClient(connection);
+
+            return _sessionProcessorCache.GetOrAdd(entityPath, (_) =>
+            {
+                ServiceBusSessionProcessor processor;
+                if (ServiceBusEntityPathHelper.ParseEntityType(entityPath) == EntityType.Topic)
+                {
+                    // entityPath for a subscription is "{TopicName}/Subscriptions/{SubscriptionName}"
+                    ServiceBusEntityPathHelper.ParseTopicAndSubscription(entityPath, out string topic, out string subscription);
+                    processor = client.CreateSessionProcessor(topic, subscription, _options.ToSessionProcessorOptions());
+                }
+                else
+                {
+                    // entityPath for a queue is "{QueueName}"
+                    processor = client.CreateSessionProcessor(entityPath, _options.ToSessionProcessorOptions());
+                }
+                processor.ProcessErrorAsync += _options.ExceptionReceivedHandler;
+                return processor;
+            });
+        }
+
+        internal ServiceBusClient CreateSessionClient(string connection) =>
+            GetOrCreateClient(connection);
+
+        private ServiceBusClient GetOrCreateClient(string connection)
+        {
+            var connectionInfo = ResolveConnectionInformation(connection);
+            if (connectionInfo.ConnectionString != null)
+            {
+                return _clientCache.GetOrAdd(
+                    connectionInfo.ConnectionString,
```

Comment:
Why don't we cache based on the `connection`?

Reply:
Do you mean the connection setting name? This is how it worked in Track 1 - caching on the connection string. I don't know if it would be possible for the connection setting to be resolved to a different value during the lifetime of the host, so I guess maybe it is safer to use the connection string?